packages/flutter/test/widgets/root_restoration_scope_test.dart

```diff
@@ -0,0 +1,400 @@
+
+  setUp(() {
+    binding._restorationManager = MockRestorationManager();
+  });
+
+  testWidgets('does not inject root bucket if inside scope', (WidgetTester tester) async {
+    final MockRestorationManager manager = MockRestorationManager();
+    final Map<String, dynamic> rawData = <String, dynamic>{};
+    final RestorationBucket root = RestorationBucket.root(manager: manager, rawData: rawData);
+    expect(rawData, isEmpty);
+
+    await tester.pumpWidget(
+      Directionality(
+        textDirection: TextDirection.ltr,
+        child: UnmanagedRestorationScope(
+          bucket: root,
+          child: const RootRestorationScope(
+            restorationId: 'root-child',
+            child: BucketSpy(
+              child: Text('Hello'),
+            ),
+          ),
+        ),
+      ),
+    );
+    manager.doSerialization();
+
+    expect(binding.restorationManager.rootBucketAccessed, 0);
+    final BucketSpyState state = tester.state(find.byType(BucketSpy));
+    expect(state.bucket.restorationId, 'root-child');
+    expect(rawData[childrenMapKey].containsKey('root-child'), isTrue);
+
+    expect(find.text('Hello'), findsOneWidget);
+  });
+
+  testWidgets('waits for root bucket', (WidgetTester tester) async {
+    final Completer<RestorationBucket> bucketCompleter = Completer<RestorationBucket>();
+    binding.restorationManager.rootBucket = bucketCompleter.future;
+
+    await tester.pumpWidget(
+      const Directionality(
+        textDirection: TextDirection.ltr,
+        child: RootRestorationScope(
+          restorationId: 'root-child',
+          child: BucketSpy(
+            child: Text('Hello'),
+          ),
+        ),
+      ),
+    );
+
+    expect(binding.restorationManager.rootBucketAccessed, 1);
+    // Child rendering is delayed until root bucket is available.
+    expect(find.text('Hello'), findsNothing);
+    expect(binding.firstFrameIsDeferred, isTrue);
+
+    // Complete the future.
+    final Map<String, dynamic> rawData = <String, dynamic>{};
+    final RestorationBucket root = RestorationBucket.root(manager: binding.restorationManager, rawData: rawData);
+    bucketCompleter.complete(root);
+    await tester.pump(const Duration(milliseconds: 100));
+
+    expect(binding.restorationManager.rootBucketAccessed, 1);
+    expect(find.text('Hello'), findsOneWidget);
+    expect(binding.firstFrameIsDeferred, isFalse);
+
+    final BucketSpyState state = tester.state(find.byType(BucketSpy));
+    expect(state.bucket.restorationId, 'root-child');
+    expect(rawData[childrenMapKey].containsKey('root-child'), isTrue);
+  });
+
+  testWidgets('no delay when root is available synchronously', (WidgetTester tester) async {
+    final Map<String, dynamic> rawData = <String, dynamic>{};
+    final RestorationBucket root = RestorationBucket.root(manager: binding.restorationManager, rawData: rawData);
+    binding.restorationManager.rootBucket = SynchronousFuture<RestorationBucket>(root);
```

Comment:
About this, it sounds like this could be a common use-case for testing. Should it be built directly inside `tester`?

Like:

```dart
tester.restorationGraph = {
  'root-child': {
    'my_id': 'Hello world',
    'sub-bucket': {
      'my_id2': 'Hello world',
    },
  },
};

await tester.pumpWidget(
  RootRestorationScope(
    restorationId: 'root-child',
    child: Column(
      children: [
         TextField(restorationId: 'my_id'),
         RestorationScope(
           restorationId: 'sub-bucket',
           child: Whatever(),
         ),
      ],
    ),
  ),
);
```

Reply:
I specifically want to avoid hand-crafted restoration data in widget tests. That data can easily change (e.g. because one of the widgets is updated with an extra RestorationScope) making these tests very brittle. Instead, you should get the widget tree into the state that you want to test and then use `tester.restartAndRestore()` (or a combination of `tester.getRestorationData()` / `tester.restoreFrom(data)`) to test that you can restore the tree from its current restoration data. That way, the test remains valid even if the restoration data format changes.

With semantics we initially made the mistake that the tests were too tightly coupled to the actual semantics tree, making them very brittle whenever we had to change the shape of the semantics tree for reasons unrelated to the actual test.

Reply:
That's a fair argument.

My fear is, by breaking this in two steps, it may cause make the test brittle because of some test that is reused between before and after the restart.
Because `restartAndRestore` doesn't fully restart the process. Singletons/Global variables are still there. And a `setup`/`tearDown` won't help, as this is withing one test, not between two tests.

If we had a `tester.setRestorationData` that was executed as the first line of a `testWidgets`, this would be compatible with `setup`/`tearDown` and avoid some hassle for users of packages like `get_it`, which is relatively popular (even if I don't like it)

Raw `provider` usage would be unaffected though.

Reply:
Would it make sense to use keys for this?
Like associate a `RestorationScope` with a `Key`, and target this bucket specifically:

```dart
RestorationScope(
  restorationId: 'foo',
  child: RestorationScope(
    key: Key('bar'),
    restorationId: 'bar',
    child: TextField(restorationId: 'id'),
  ),
);

...

tester.setRestorationDataFor(find.byKey(Key('bar')), { 'id': 'value' });
```

This would avoid having to specify the entire object and allows targeting a very specific scope – which should make the issue you mentioned less problematic